<a href="https://colab.research.google.com/github/JakeC05/Engineering-Methods/blob/main/lab2_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
from sympy import Matrix, symbols
import numpy as np

# **Problem 1: Finding Eigenvalues and Eigenvectors for Diagonalization**

The goal of this problem is to compute the eigenvalues/eigenvectors of a matrix, then diagonalize when possible.

In [2]:
# Define the matrix
M = Matrix([[2, 1, 0, 0],
    [1, 2, 0, 0],
    [0, 0, 3, 1],
    [0, 0, 1, 3]])

# Display
M

Matrix([
[2, 1, 0, 0],
[1, 2, 0, 0],
[0, 0, 3, 1],
[0, 0, 1, 3]])

## **(a) Characteristic Polynomial**
To find the characteristic polynomial of $M$, we need to instantiate an algebraic variable: we do this by using `symbols()`. For example, if we want x to represent an algebraic variable, then we write `x = symbols('x')` (notice the quotations on the inside, this data-structure is known as a **string**). Then the characteristic polynomial can be written as `char_poly = (M-x*I).det()`, where the identity matrix $I$ is given by `Matrix.eye(4)`

**Create the characteristic polynomial under `char_poly` of the matrix $M$.**


In [4]:
# Create variable
x=symbols('x')
# Characteristic polynomial
I = Matrix.eye(4) # 4 since A is a 4x4 matrix
char_poly=(M-x*I).det()
#display
char_poly

(x**2 - 6*x + 8)*(x**2 - 4*x + 3)

## **(b) Eigenvalues**
For a matrix $M$, then `M.eigvals()` returns a set of the form $\{\text{eigenvalues of A}:\text{algebraic multiplicity}\}$, where the **algebraic multiplicity** of an eigenvalue $\lambda$ is the maximum integer $k$ such that $(x-\lambda)^k$ divides the characteristic polynomial $\det(M-xI)$.

**Compute the eigenenvalues of $M$. From this, deduce whether $M$ is diagonalizable.**

In [6]:
# Compute eigenvalues here
M.eigenvals()


{3: 1, 1: 1, 4: 1, 2: 1}

Answer: 1,2,3,4 all multiplicity of 1

## **(c) Eigenvectors**

SymPy can find eigenvectors with `M.eigenvects()`.

1. Use `M.eigenvects()` to see the eigenvalues, their algebraic multiplicities,
   and bases for each eigenspace $E_\lambda$.
2. For each eigenvalue, pick one eigenvector and write it down.
3. For one eigenpair $(\lambda, v)$, verify by hand (or in code) that
   $$
   M v = \lambda v.
   $$


In [7]:
#1.
M.eigenvects()

[(1,
  1,
  [Matrix([
   [-1],
   [ 1],
   [ 0],
   [ 0]])]),
 (2,
  1,
  [Matrix([
   [ 0],
   [ 0],
   [-1],
   [ 1]])]),
 (3,
  1,
  [Matrix([
   [1],
   [1],
   [0],
   [0]])]),
 (4,
  1,
  [Matrix([
   [0],
   [0],
   [1],
   [1]])])]

2,3:
Did $Av = \lambda v$?

Answer (Binary):

## **(d) Diagonalization**

A matrix $M$ is **diagonalizable** if it is similar to diagonal matrix $D$: that is, if we can find an invertible matrix $P$ and a diagonal matrix $D$
such that
$$
M = P D P^{-1},
$$
where the columns of $P$ are eigenvectors of $M$ and the diagonal entries of $D$ are the eigenvalues.

1. Use `M.diagonalize()` to compute matrices $P$ and $D$.
- This gives a tuple, so you want `P,D = M.diagonalize()`
2. Check in code that `P.inv()*M*P` really equals $D$.
3. Why does having 4 linearly independent eigenvectors imply that $M$ is diagonalizable?

In [8]:
# 1. Compute P and D

P,D=M.diagonalize()

display (P)


Matrix([
[-1,  0, 1, 0],
[ 1,  0, 1, 0],
[ 0, -1, 0, 1],
[ 0,  1, 0, 1]])

In [9]:
display (D)


Matrix([
[1, 0, 0, 0],
[0, 2, 0, 0],
[0, 0, 3, 0],
[0, 0, 0, 4]])

In [10]:
# 2. Check whether P.inv()*M*P == D is true
P.inv()*M*P==D

True

Answer to 3: This let's us build the matrix $P$. If we did not have enough eigenvectors, then we would not be able to build $P$: any collection of eigenvectors of $M$ would build a singular matrix $P$.

# Problem 2 — Audience Attention in a 11-Actor Play

Eleven actors stand in a line on stage. At each "beat" of a performance,
the audience's attention shifts between them according to simple rules:

- Actors **1** and **11** (edges) keep 60% of their attention and send 40% to their single neighbor.
- Actors **2-5** and **7-10** (non-center,non-edge) keep 60% of attention, sends 20% to the neighbor on the left, and 20% to the neighbor on the right.
- Actor **6** keeps 85% of attention, sends 7.5% to Actor **10**, and 7.5% to Actor **12**.


Let $x_k$ be the vector representing the attention vector after the $k$-th beat.

The update rule is linear: $x_{k+1} = A x_k$,
where $A$ is the matrix

$$
A =
\begin{bmatrix}
0.6 & 0.2 & 0.0 & 0.0 & 0.0 & 0.0  & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.4 & 0.6 & 0.2 & 0.0 & 0.0 & 0.0  & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.2 & 0.6 & 0.2 & 0.0 & 0.0  & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.2 & 0.6 & 0.2 & 0.0  & 0.0 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.2 & 0.6 & 0.075& 0.0 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.2 & 0.85 & 0.2 & 0.0 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.075& 0.6 & 0.2 & 0.0 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0  & 0.2 & 0.6 & 0.2 & 0.0 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0  & 0.0 & 0.2 & 0.6 & 0.2 & 0.0 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0  & 0.0 & 0.0 & 0.2 & 0.6 & 0.4 \\
0.0 & 0.0 & 0.0 & 0.0 & 0.0 & 0.0  & 0.0 & 0.0 & 0.0 & 0.2 & 0.6
\end{bmatrix}.
$$

We call $A$ the ***attention-flow matrix***.

Famous actor Bedro Bascall is in the center of the line-up as actor 6, so we begin with all attention focused on Actor 6:
$$
x_0 :=   [0,0,0,0,0,100,0,0,0,0,0]^T.
$$

Thus $x_k := A^kx_0$ measures the exact amount of attention each actor has after k beats, assuming the play begins with the audience looking entirely at Actor 6.


We will:
1. Simulate this system numerically using NumPy.  
2. Diagonalize the matrix.  
3. Interpret the long-term behavior.

In [12]:
# Load A

A =  Matrix([
    [0.6, 0.2, 0.0, 0.0,   0.0,   0.0,   0.0,  0.0,  0.0,  0.0,  0.0],
    [0.4, 0.6, 0.2, 0.0,   0.0,   0.0,   0.0,  0.0,  0.0,  0.0,  0.0],
    [0.0, 0.2, 0.6, 0.2,   0.0,   0.0,   0.0,  0.0,  0.0,  0.0,  0.0],
    [0.0, 0.0, 0.2, 0.6,   0.2,   0.0,   0.0,  0.0,  0.0,  0.0,  0.0],
    [0.0, 0.0, 0.0, 0.2,   0.6,   0.075, 0.0,  0.0,  0.0,  0.0,  0.0],
    [0.0, 0.0, 0.0, 0.0,   0.2,   0.85,  0.2,  0.0,  0.0,  0.0,  0.0],
    [0.0, 0.0, 0.0, 0.0,   0.0,   0.075, 0.6,  0.2,  0.0,  0.0,  0.0],
    [0.0, 0.0, 0.0, 0.0,   0.0,   0.0,   0.2,  0.6,  0.2,  0.0,  0.0],
    [0.0, 0.0, 0.0, 0.0,   0.0,   0.0,   0.0,  0.2,  0.6,  0.2,  0.0],
    [0.0, 0.0, 0.0, 0.0,   0.0,   0.0,   0.0,  0.0,  0.2,  0.6,  0.4],
    [0.0, 0.0, 0.0, 0.0,   0.0,   0.0,   0.0,  0.0,  0.0,  0.2,  0.6],
])
# Load x0
x0 = Matrix([
    0,
    0,
    0,
    0,
    0,
    100,  # Bedro Bascal (actor 6)
    0,
    0,
    0,
    0,
    0
])

A

Matrix([
[0.6, 0.2, 0.0, 0.0, 0.0,   0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
[0.4, 0.6, 0.2, 0.0, 0.0,   0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
[0.0, 0.2, 0.6, 0.2, 0.0,   0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
[0.0, 0.0, 0.2, 0.6, 0.2,   0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
[0.0, 0.0, 0.0, 0.2, 0.6, 0.075, 0.0, 0.0, 0.0, 0.0, 0.0],
[0.0, 0.0, 0.0, 0.0, 0.2,  0.85, 0.2, 0.0, 0.0, 0.0, 0.0],
[0.0, 0.0, 0.0, 0.0, 0.0, 0.075, 0.6, 0.2, 0.0, 0.0, 0.0],
[0.0, 0.0, 0.0, 0.0, 0.0,   0.0, 0.2, 0.6, 0.2, 0.0, 0.0],
[0.0, 0.0, 0.0, 0.0, 0.0,   0.0, 0.0, 0.2, 0.6, 0.2, 0.0],
[0.0, 0.0, 0.0, 0.0, 0.0,   0.0, 0.0, 0.0, 0.2, 0.6, 0.4],
[0.0, 0.0, 0.0, 0.0, 0.0,   0.0, 0.0, 0.0, 0.0, 0.2, 0.6]])

## **(a) Applying the matrix several times**

Compute $x_1 = A x_0$, $x_{10} = A^{10} x_0$, and $x_{1000} = A^{1000} x_0$.


In [13]:
# compute x1
x1=A*x0

In [14]:
# compute x10
x10=A**10*x0

In [15]:
# compute x1000
x1000=A**1000*x0

## **(b) Intrepretation**
  Look at the entries of the vectors from (a):
   1. How does attention spread from Actor 6 to the others?
   2. Does the **total** attention (sum of all 11 entries of $x_k$) change? (You can compute this with `sum(x1)`)

Answer 1: Attention spreads from actor 3 outward. After a few beats, attention begins spread out to the other actors. It seems like the vector converges to a vector somewhat of the form $$[4.29,8.57,\cdots, 8.57,4.29 ]^T$$

In [17]:
# 2. Compute the sum of the entries of x1, x5, and x10: compute using sum()
print(sum(x1))
print(sum(x10))
print(sum(x1000))

100.000000000000
100.000000000000
100.000000000001


Answer 2: They all add to 100. Notice that the last entry has computational error from the computer.

## **(c) Diagonalization and long-term behavior**

Recall: SymPy can diagonalize $A$ with `A.diagonalize()`.

Compute matrices $P$ and $D$ such that
   $$
   A = P D P^{-1}.
   $$



In [20]:
# Diagonalize A: construct P, D.
P,D=A.diagonalize()

display (D)


Matrix([
[0.217430396420866,                 0,                0,                 0,                 0,   0,                 0,                 0,   0,                 0,                 0],
[                0, 0.219577393481939,                0,                 0,                 0,   0,                 0,                 0,   0,                 0,                 0],
[                0,                 0, 0.34728727110711,                 0,                 0,   0,                 0,                 0,   0,                 0,                 0],
[                0,                 0,                0, 0.364885899083011,                 0,   0,                 0,                 0,   0,                 0,                 0],
[                0,                 0,                0,                 0, 0.560264290747211,   0,                 0,                 0,   0,                 0,                 0],
[                0,                 0,                0,                 0,      

# (c) Approximating $\lim_{k\rightarrow \infty}A^kx_0$
We want to approximate the above limit by setting $k$ to be a gigantic number, but this is computationally expensive. Thus we will use $A=PDP^{-1}$ and take high powers of the left-hand side.

1. Compute $PD^{\text{big_num}}P^{-1}$, where big_num $= 10^{50}$.
2. Approximate $\lim_{k\rightarrow \infty}x_{k}$.

In [32]:
#1
# construct big_n

# Compute PD^(big_n)P^{-1}
big_num=10**50

P*D**(big_num)*P.inv()


Matrix([
[0.0428571428571429, 0.0428571428571429, 0.0428571428571429, 0.0428571428571429, 0.0428571428571429, 0.0428571428571428, 0.0428571428571428, 0.0428571428571429, 0.0428571428571429, 0.0428571428571429, 0.0428571428571429],
[0.0857142857142858, 0.0857142857142858, 0.0857142857142858, 0.0857142857142858, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142858, 0.0857142857142858, 0.0857142857142858],
[0.0857142857142858, 0.0857142857142858, 0.0857142857142858, 0.0857142857142858, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142858, 0.0857142857142858, 0.0857142857142858],
[0.0857142857142857, 0.0857142857142858, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.0857142857142858, 0.0857142857142858],
[0.0857142857142857, 0.0857142857142858, 0.0857142857142857, 0.0857142857142857, 0.0857142857142857, 0.

Answer 2: each column converges to:
        [0.0428
          0.0857
          ...
          0.228
          ...
          0.0857
          0.0428]

## **(d) Interpret limits**

Explain your findings in part (c) in terms of our play.

Answer: after a long time this is where the attention will be split

## **(e) Interpret eigenvectors**
What do eigenvectors represent in our scenario?

Answer: they represnt where the attention will go after a long time has passed